### ISSUE TITLE
Losing saved session

### ISSUE THREAD

**Issue by murilosr** [Adamcake/Bolt#84]
I'm using flatpak version of Bolt on PopOS 22.

Sometimes when I start the launcher, the session is saved and I can start the game normally.
But sometimes when the launcher is started, it shows the license notice shows (AGPL-3.0, etc) and the characters are not loaded. I need to login again on Jagex account to launch the client.

What happens to lose the session? Because it seems random. Is it normal?

**Comment by Adamcake** [comment]
The licence and disclaimer gets shown if your credentials file has no sessions stored in it - not even expired ones. The credentials file is in XDG_DATA_HOME, for flatpak it would be `~/.var/app/com.adamcake.Bolt/data/bolt-launcher/creds`.

A possible explanation would be that the location of Bolt's data folder is changing, but with flatpak that seems unlikely. It's possible that Bolt's UI is overwriting that file with an empty list but I can't think of a reason why it would do that. Maybe if you're opening it, launching the game and closing the launcher in very quick succession? But even if that bug existed then I'd think it would save expired credentials not delete them completely.

**Comment by murilosr** [comment]
No, this happens often overnight. How could I debug it? Or do you suggest something to try?

**Comment by Adamcake** [comment]
Files don't change themselves overnight. Whatever's happening, it most likely happens when opening the launcher. You'd need to monitor that `creds` file to find out what the contents are just before and after you open the launcher and get the licence message. If it's really empty, then when is it getting emptied, and why?

**Comment by murilosr** [comment]
Yeah, files don't change alone... I'm observing what is happening to the system, and the "creds" file is cleaned as soon as I start the launcher.

I made a backup of the content (creds_backup) just before opening, and it got cleaned:

**Before:**
```
cat creds_backup 
[{"user":{"id":"66wiD***","userId":"7f4t***","displayName":"Murilo","suffix":"****"},"accounts":[{"accountId":"179***",...
```

**After**
```
cat creds
[]
```

I suspected it could be after the system is turned off and on, but today, I played around 8:55 AM. Now it is 01:55 PM, and I haven't restarted the system, and the credentials were lost, as I'm showing.

----

> If it's really empty, then when is it getting emptied, and why?

So the when you have the answer: When the launcher is started.
How can I get the why?

**Comment by murilosr** [comment]
Since I copied the file before opening the launcher to the 'creds_backup' file, I just reversed the backup as 'creds' file and when I opened the launcher it was with my session.

**Comment by Adamcake** [comment]
So the file isn't getting deleted, but getting overwritten with `[]` when you start the launcher? But only sometimes?

Maybe it's failing to open/read the file and assuming it doesn't exist. Does it print any errors in a terminal?

**Comment by Adamcake** [comment]
It might actually be due to you having the file open in an editor, if the editor keeps the file handle open?

**Comment by murilosr** [comment]
> So the file isn't getting deleted, but getting overwritten with `[]` when you start the launcher? But only sometimes?
> 
> Maybe it's failing to open/read the file and assuming it doesn't exist. Does it print any errors in a terminal?

Yes, it's overwriting to []. It's sometimes. I'm launching via desktop entry, but I'll start using the terminal way to see if it prints any error.

> It might actually be due to you having the file open in an editor, if the editor keeps the file handle open?

No, the file is not opened in any editor, I just printed and handled it via terminal (cat, cp and mv commands only).

**Comment by murilosr** [comment]
The content of the .desktop file I'm using:
(But I'm pretty sure that this doesn't influence it)

~/.local/share/applications/com.adamcake.Bolt.RuneScape.desktop 

```
[Desktop Entry]
Type=Application
Name=RuneScape
GenericName=RuneScape
Icon=runescape
Terminal=false
NoDisplay=true
StartupWMClass=RuneScape
X-Flatpak=com.adamcake.Bolt
Exec=/usr/bin/flatpak run --branch=stable --arch=x86_64 com.adamcake.Bolt
```

**Comment by murilosr** [comment]
It just happened using the terminal run from flatpak. This is the complete log until after the religion to Jagex Account:


```
murilo@murilo-avell :: ~ » flatpak run com.adamcake.Bolt 
[1113/132029.228233:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
Gtk-Message: 13:20:29.319: Failed to load module "canberra-gtk-module"
Gtk-Message: 13:20:29.322: Failed to load module "canberra-gtk-module"
[1113/132029.368134:WARNING:chrome_browser_cloud_management_controller.cc(87)] Could not create policy manager as CBCM is not enabled.
[B] OnContextInitialized
[B] Browser::Window popup constructor, this=99031499679120
[B] OnWindowCreated 0 this=99031499679120
[1113/132029.472430:WARNING:key_storage_linux.cc(183)] OSCrypt tried Libsecret but couldn't initialise.
[B] Browser::OnAfterCreated for browser 1
[B] OnBrowserCreated this=99031499679120 1
[R] OnBrowserCreated for browser 1
[R] OnLoadEnd for browser 1
[B] Browser::OnBeforePopup for browser 1
[B] GetDelegateForPopupBrowserView this=99031499679120
[B] Browser::Window popup constructor, this=99031499865104
[B] OnPopupBrowserViewCreated this=99031499679120
[B] OnWindowCreated 0 this=99031499865104
[B] Browser::OnAfterCreated for browser 2
[B] OnBrowserCreated this=99031499865104 2
[R] OnBrowserCreated for browser 2
[R] OnBrowserCreated for browser 2
[1113/132132.243211:ERROR:ev_root_ca_metadata.cc(161)] Failed to decode OID: 0
[R] OnBrowserCreated for browser 2
[R] OnLoadEnd for browser 2
[1113/132132.761242:WARNING:browser_info.cc(309)] Returning a speculative frame for 34359738376 [8,8]
[R] OnBrowserCreated for browser 2
[1113/132132.829170:WARNING:browser_info.cc(309)] Returning a speculative frame for 38654705672 [9,8]
[1113/132132.829351:WARNING:browser_info.cc(309)] Returning a speculative frame for 38654705672 [9,8]
[1113/132132.829400:WARNING:browser_info.cc(309)] Returning a speculative frame for 38654705672 [9,8]
[1113/132132.829432:WARNING:browser_info.cc(309)] Returning a speculative frame for 38654705672 [9,8]
[1113/132132.832165:WARNING:browser_info.cc(309)] Returning a speculative frame for 38654705672 [9,8]
[R] OnBrowserCreated for browser 2
[R] OnLoadEnd for browser 2
[R] OnLoadEnd for browser 2
[R] OnLoadEnd for browser 2
[R] OnLoadEnd for browser 2
[R] OnBrowserCreated for browser 2
[R] OnLoadEnd for browser 2
[R] OnLoadEnd for browser 2
[B] CanClose for window 0, this=99031499865104
[B] OnWindowClosing 0 this=99031499865104
[B] OnBrowserDestroyed this=99031499865104
[B] Browser::OnBeforeClose for browser 2
```


This is a log running again, but without losing the session:
```
murilo@murilo-avell :: ~ » flatpak run com.adamcake.Bolt
[1113/132352.487465:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
Gtk-Message: 13:23:52.538: Failed to load module "canberra-gtk-module"
Gtk-Message: 13:23:52.539: Failed to load module "canberra-gtk-module"
[1113/132352.552221:WARNING:chrome_browser_cloud_management_controller.cc(87)] Could not create policy manager as CBCM is not enabled.
[B] OnContextInitialized
[B] Browser::Window popup constructor, this=105553562546144
[B] OnWindowCreated 0 this=105553562546144
[1113/132352.561836:WARNING:key_storage_linux.cc(183)] OSCrypt tried Libsecret but couldn't initialise.
[B] Browser::OnAfterCreated for browser 1
[B] OnBrowserCreated this=105553562546144 1
[R] OnBrowserCreated for browser 1
[R] OnLoadEnd for browser 1
[1113/132352.720097:ERROR:ev_root_ca_metadata.cc(161)] Failed to decode OID: 0
```

I checked again and the 'creds' file was overwritten to '[]' just after the launch (I used 'cat' only, not text editor or something else that kept the file open).
Before that, I played other 3 times with no session loss. The 'creds' file was checked before launch just like I did now, so that's not related.

**Comment by Adamcake** [comment]
The only thing I can think of that could possibly be happening is that the file is failing to load here:

https://github.com/Adamcake/Bolt/blob/master/src/browser/window_launcher.cxx#L488-L496

If that fails for any reason (such as not existing), a default file is written with `[]` in it. That seems to be the only thing that could possibly be happening here.

But why is it failing? I'm not sure. A tool like `strace` would tell you, if you run `strace flatpak run com.adamcake.Bolt` in a terminal, but you'll have to search a lot of output from the Chromium engine to find it. Here's an example of what you're looking for.
![image](https://github.com/user-attachments/assets/80be14cb-2cb1-42dd-839d-08c9174742b5)

**Comment by murilosr** [comment]
Using the strace like you said, just after the launcher opened it showed a code 400 error in the log session of the screen with the message:

**Discarding expired login, status: 400. Please sign in again.**

Searching in the code base, it would be [this line](https://github.com/Adamcake/Bolt/blob/65a26a1bb4d8f1366c81170a097b7604823bb1bb/app/src/main.ts#L166)

The difference this time (to the other reported ones) is that it displayed this message in red at the bottom of the window. After I closed the launcher and opened it again, the License Notice popup showed, like the other reported times (they didn't even show that the session expired).

It really seems to be related to the refresh OAuth Session, because it seems to occur 24h after my last valid login (that's why initially I said somehow it happened "overnight").

I have the original strace file, if you want I can send it to you privately so you can debug better (9.5MB).

**Comment by Adamcake** [comment]
That's what I would've expected to happen if the credentials were actually expired or invalid. The only reason that it would discard them, without printing any errors at all, is if it didn't think it had any credentials to begin with. This error is probably unrelated to your other one, I think - it would have been easy to cause it by managing your creds and creds_backup files incorrectly, like restoring an old backup that you already used once before, for example.

**Comment by machadofguilherme** [comment]
I am going through the same problem. Occurs from time to time.

**Comment by murilosr** [comment]
I still have the logoffs daily, as I reported since the opening message here.

@Adamcake thinks I'm messing with the creds file, keeping it opened in the editor, etc., but the error started before I even knew about its existence. Since Nov. 14th I haven't touched it anymore and still get logged off.

I'll try to get more information and debug it myself, because I'm not being believed.

@machadofguilherme, what is your setup? (OS? using flatpak or built yourself? Any other infos?)

**Comment by machadofguilherme** [comment]
@murilosr @Adamcake 
First of all, I think you are BR. I use NixOS as a system and I installed Bolt through Flatpak. I did not observe any logs, just thought that the fix would come soon and that the developer was aware of this bug.

**Comment by Adamcake** [comment]
It's not that I don't believe you, I'm just struggling to think of scenarios where what you're describing would happen. I think the most recent case, where you got a message about expired credentials, is most likely unrelated to the other issue, and I suggest that it may have been caused by your creds_backup system.

If the original issue is caused by Bolt failing to read the creds file - which I think is the only possibility - then `strace -f` would tell you the reason.

**Comment by kmcgurty** [comment]
According to [this article](https://www.oauth.com/oauth2-servers/access-tokens/access-token-response), a 400 response could be for any reason, but will include a reason for why it failed. In the the [AuthService.refreshOAuthToken](https://github.com/Adamcake/Bolt/blob/master/app/src/lib/Services/AuthService.ts#L98) function, we discard the body of the request, which could have valuable info as to what's going on.

`refreshOAuthToken` could be changed to return a `Promise<Result<AuthTokens, string>>`, and instead of returning `xml.status`, it could return `xml.responseText`, or parse `responseText` as JSON, and return the description string if it exists.

**Comment by doorknob60** [comment]
For what it's worth, I've also been logged out a couple times since I started using Bolt about a week ago (just switched over to a Jagex account). I haven't been able to track down any patterns yet, or do any debugging. I do believe something may be going on, but I'm not sure what yet. But I will try to keep an eye on any patterns, and the status of the creds file, in case I discover anything useful. I'm using Arch with the Flatpak.

**Comment by murilosr** [comment]
I'm still trying to find out what make it happens. I'm keeping a log when it happens, and some days it happened twice, so it is not time-related.

I'm suspecting that make it "happens often" when the launcher is started and almost immediately I click the "Play button" (because the character was already selected), before the log message "Signed into saved user 'Murilo'" appears. I'm using the launcher settings to close after launching the client.

Do you think this could be related? Maybe the launcher is in the middle of refreshing the auth session, and when it closes it loses the creds information, or something like that.

Like I said, I'm suspecting that something could be related to this because when I take some time to click the Play Button (and the "signed into..." message shows), it appears that I do not lose the session (I kept 3 days straight without losing session). I couldn't reproduce the losing session error even once while using strace tool, just the "Discarding expired login" error that I already registered here.

**Comment by Adamcake** [comment]
That's possible in theory. Someone else on Discord has the same issue, I got them to run an strace and the file is definitely loading correctly. The UI would never show a disclaimer if it receives any non-empty list of credentials at all, but maybe some action you're doing is causing it to save an empty list and exit immediately, which means the disclaimer gets shown next time you run it.

@kmcgurty you're the expert here, what do you think?

**Comment by murilosr** [comment]
For the first time I could record the start of launcher losing the session with strace, but it doesn't seem to help.

![Screenshot from 2024-11-27 08-20-07](https://github.com/user-attachments/assets/323d9e17-0f58-48b9-bf05-10690dac3cbd)

![Screenshot from 2024-11-27 08-20-59](https://github.com/user-attachments/assets/f6c882aa-c270-4e7e-b6a7-a34ef16f0ce6)

It seems that the write of '[]' was after the read (at least in the order that appears on the Strace log).
This time, I didn't see the content of the credentials before running, but other times, I already confirmed that the session existed and was lost after the application launched (more than 3x this was confirmed).

Given that, could it be a race condition that halts the reading because it if first writing '[]' and then it reads? Because it seems that they are in different processes (8421 and 8446). I'll attach the whole log here, it seems it doesn't have anything particular about my session.

Log file: https://drive.google.com/file/d/1aOiYtXUfUpoOD-AQe9D4yTF9k2uml9rS/view?usp=sharing

**Comment by whitelje** [comment]
@murilosr @doorknob60 
Do you play RS on any other devices?

**Comment by doorknob60** [comment]
> @murilosr @doorknob60 Do you play RS on any other devices?

I do, but I noticed occasional logging out even when I was only using one device. It doesn't seem to affect it in my limited experience. All that being said, most of the time the past week it is staying logged in for me. I did turn off the "Close Bolt after launching a game option" a few days ago, not sure if it has made a difference.

**Comment by murilosr** [comment]
> @murilosr @doorknob60 Do you play RS on any other devices?

Yes, on mobile.

**Comment by murilosr** [comment]
At least for me, it's confirmed that if I wait for the log "Signed into saved user 'Murilo'" to appear to click the "Play" button, my session is not lost anymore.
The last time it happened was when I captured the lost session with strace that I posted here (8 days ago). Before that, I lost every day (sometimes more than one time on the same day)

**Comment by Jookia** [comment]
I can confirm that clicking play before the signed in message comes up wipes my creds file.
